### DISCUSSION TITLE
Widget animation slowing app

### DISCUSSION THREAD

**Discussion by akhudek** [fltk-rs/fltk-rs#1361] [Q&A] [answered]
I have a widget that I update at 60fps via a separate thread and using app::awake(). However, I've noticed that simply calling awake() alone without any actual drawing code causes a slight lag when typing in the editor component, even though I'm only calling redraw on a specific widget.

Is there something I'm doing wrong that is causing it to redraw the whole window rather than just the widget? I call redraw on the specific widget only then call awake().

**Comment by MoAlyousef** [comment, marked as answer]
Hi
Can you create a minimal repro which shows the problem.
Do you have any redraw calls? Sometimes they're needed when for example changing something in the widget causes damage to the parent or with resizing, so redrawing the parent is necessary. In that case you can place the widget inside a sub-parent (if its direct parent was the window), that way you avoid redrawing the whole window.
Sometimes it isn't necessary:
```rust
use fltk::{prelude::*, *,};

fn main() {
    let app = app::App::default().with_scheme(app::Scheme::Oxy);
	let buf = text::TextBuffer::default();
    let mut win = window::Window::default().with_size(400, 300);
    let mut ed = text::TextEditor::default().with_size(400, 200);
	ed.set_buffer(buf);
	let mut f = frame::Frame::default().with_size(200, 40).below_of(&ed, 5).center_x(&ed);
	win.end();
    win.show();
	
	std::thread::spawn(move || {
		loop {
			f.set_label(&format!("{:?}", std::time::Instant::now()));
			app::awake();
			std::thread::sleep(std::time::Duration::from_millis(16));	
		}
	});
    
	app.run().unwrap();
}
```
This for example adds a group as a parent to our previous frame:
```rust
use fltk::{prelude::*, *,};

fn main() {
    let app = app::App::default().with_scheme(app::Scheme::Oxy);
	let buf = text::TextBuffer::default();
    let mut win = window::Window::default().with_size(400, 300);
    let mut ed = text::TextEditor::default().with_size(400, 200);
	ed.set_buffer(buf);
	let mut grp = group::Group::default().with_size(200, 40).below_of(&ed, 20).center_x(&ed);
	let mut f = frame::Frame::default().with_size(200, 40).center_of_parent();
	f.set_frame(enums::FrameType::FlatBox);
	grp.end();
	win.end();
    win.show();
	
	let mut i: u8 = 0;
	std::thread::spawn(move || {
		loop {
			f.set_color(enums::Color::by_index(i));
			if i % 2 == 0 {
				f.resize(f.x(), f.y(), 100, 40);
			} else {
				f.resize(f.x(), f.y(), 200, 40);
			}		
			f.parent().unwrap().redraw();
			app::awake();
			std::thread::sleep(std::time::Duration::from_millis(16));	
			i+=1;
			if i == 255 {
				i = 0;
			}
		}
	});
    
	app.run().unwrap();
}
```

**Reply by akhudek** [reply]
Ah, this is helpful. In trying to create a smaller example I realized that it was my event loop causing the slowdown. I was checking a crossbeam channel but of course there isn't always an event message when you use awake(). Switching to try_recv() fixed it. 

Thanks!
